### PR TITLE
Added dynamic tables for CRDs & fixed aggregator functionality for dynamic tables

### PR DESCRIPTION
**Pull request by bigdatasourav**
- [x] Aggregator functionality is working
- [x] There should not be any conflict between columns of the custom resource tables

### CODE REVIEW

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+                  type: string
+                privateKey:
+                  description: Options to control private keys used for the Certificate.
+                  type: object
+                  properties:
+                    algorithm:
+                      description: Algorithm is the private key algorithm of the corresponding private key for this certificate. If provided, allowed values are either `RSA`,`Ed25519` or `ECDSA` If `algorithm` is specified and `size` is not provided, key size of 256 will be used for `ECDSA` key algorithm and key size of 2048 will be used for `RSA` key algorithm. key size is ignored when using the `Ed25519` key algorithm.
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:145]:
```suggestion
                      description: Algorithm is the private key algorithm of the corresponding private key for this certificate. If provided, allowed values are either `RSA`, `Ed25519`, or `ECDSA` If `algorithm` is specified and `size` is not provided, a key size of 256 will be used for the `ECDSA` key algorithm and key size of 2048 will be used for `RSA` key algorithm. key size is ignored when using the `ED25519` key algorithm.
```

Can we also mention the key size units in the description? Like 256 bits/bytes

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+                        - ECDSA
+                        - Ed25519
+                    encoding:
+                      description: The private key cryptography standards (PKCS) encoding for this certificate's private key to be encoded in. If provided, allowed values are `PKCS1` and `PKCS8` standing for PKCS#1 and PKCS#8, respectively. Defaults to `PKCS1` if not specified.
+                      type: string
+                      enum:
+                        - PKCS1
+                        - PKCS8
+                    rotationPolicy:
+                      description: RotationPolicy controls how private keys should be regenerated when a re-issuance is being processed. If set to Never, a private key will only be generated if one does not already exist in the target `spec.secretName`. If one does exists but it does not have the correct algorithm or size, a warning will be raised to await user intervention. If set to Always, a private key matching the specified requirements will be generated whenever a re-issuance occurs. Default is 'Never' for backward compatibility.
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:158]:
```suggestion
                      description: RotationPolicy controls how private keys should be regenerated when a re-issuance is being processed. If set to Never, a private key will only be generated if one does not already exist in the target `spec.secretName`. If one does exist but does not have the correct algorithm or size, a warning will be raised to await user intervention. If set to Always, a private key matching the specified requirements will be generated whenever a re-issuance occurs. The default is 'Never' for backward compatibility.
```

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+                    type: string
+            status:
+              description: Status of the Certificate. This is set and managed automatically.
+              type: object
+              properties:
+                conditions:
+                  description: List of status conditions to indicate the status of certificates. Known condition types are `Ready` and `Issuing`.
+                  type: array
+                  items:
+                    description: CertificateCondition contains condition information for an Certificate.
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:209]:
```suggestion
                    description: CertificateCondition contains condition information for a Certificate.
```

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+                        type: string
+                      status:
+                        description: Status of the condition, one of (`True`, `False`, `Unknown`).
+                        type: string
+                        enum:
+                          - "True"
+                          - "False"
+                          - Unknown
+                      type:
+                        description: Type of the condition, known values are (`Ready`, `Issuing`).
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:237]:
```suggestion
                        description: Type of the condition, known values are `Ready` and `Issuing`.
```

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
++------------------------------------+--------------------------------------+-------------+--------------------+-----------+
+```
+
+## Examples
+
+### Inspect the table structure
+
+List all tables:
+
+```sql
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:304]:
```suggestion
```bash
```

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+| kubernetes_cluster_role               | ClusterRole contains rules that represent a set of permissions.                                                                                                  |
+| kubernetes_cluster_role_binding       | A ClusterRoleBinding grants the permissions defined in a cluster role to a user or set of users. Access granted by ClusterRoleBinding is cluster-wide.           |
+| ...                                   | ...                                                                                                                                                              |
++---------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+To get details of a specific custom resource table, inspect it by name:
+
+```sql
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:318]:
```suggestion
```bash
```

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+spec:
+  secretName: temporal-w-spcloud123456-tls
+  duration: 87600h # 10 years
+  dnsNames:
+    - w-spcloud123456
+```
+
+This plugin will automatically create a table called `kubernetes_certificate`:
+
+```
```
Comment (misraved) [inline docs/tables/kubernetes_{custom_resource_singular_name}.md:289]:
```suggestion
```bash
```

File: kubernetes/plugin.go [hunk 8]
```diff
@@ -73,66 +77,77 @@ func pluginTableDefinitions(ctx context.Context, d *plugin.TableMapData) (map[st
...
+					ctx = context.WithValue(ctx, contextKey("VersionSchemaSpec"), version.Schema.OpenAPIV3Schema.Properties["spec"])
+					ctx = context.WithValue(ctx, contextKey("VersionSchemaStatus"), version.Schema.OpenAPIV3Schema.Properties["status"])
+					if len(version.Schema.OpenAPIV3Schema.Description) > 0 {
+						ctx = context.WithValue(ctx, contextKey("VersionSchemaDescription"), strings.TrimSuffix(version.Schema.OpenAPIV3Schema.Description, ".")+".")
+					}
+				}
+			}
+		}
+
+		// add the tables in sneak case
```
Comment (misraved) [inline kubernetes/plugin.go:103]:
```suggestion
		// add the tables in snake case
```

File: kubernetes/plugin.go [hunk 9]
```diff
@@ -73,66 +77,77 @@ func pluginTableDefinitions(ctx context.Context, d *plugin.TableMapData) (map[st
...
+	if err != nil {
+		plugin.Logger(ctx).Error("listK8sDynamicCRDs", "GetNewClientCRDRaw", err)
+		return nil, err
+	}
+
+	input := metav1.ListOptions{
+		Limit:          500,
+		TimeoutSeconds: types.Int64(5),
+	}
+
```
Comment (misraved) [inline kubernetes/plugin.go:128]: Should we add code to check if the user specifies a limit lower than the hard limit of `500`?

File: kubernetes/plugin.go [hunk 10]
```diff
@@ -73,66 +77,77 @@ func pluginTableDefinitions(ctx context.Context, d *plugin.TableMapData) (map[st
...
+	crds := []v1.CustomResourceDefinition{}
+
+	pageLeft := true
+	for pageLeft {
+		response, err := clientset.ApiextensionsV1().CustomResourceDefinitions().List(ctx, input)
+		if err != nil {
+			// Handle err at the plugin load time if config is not setup properly
+			if strings.Contains(err.Error(), "/apis/apiextensions.k8s.io/v1/customresourcedefinitions?limit=500") {
```
Comment (misraved) [inline kubernetes/plugin.go:136]: - [ ] Can we add a better comment describing the error that we are handling here?

File: kubernetes/table_kubernetes_custom_resource.go [hunk 11]
```diff
@@ -1,120 +1,209 @@
...
+
+		resourceId := schema.GroupVersionResource{
+			Group:    groupName,
+			Version:  activeVersion,
+			Resource: resourceName,
+		}
+
+		response, err := clientset.Resource(resourceId).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			if strings.Contains(err.Error(), "could not find the requested resource") {
```
Comment (misraved) [inline kubernetes/table_kubernetes_custom_resource.go:135]:
```suggestion
                        // Handle not found error code
			if strings.Contains(err.Error(), "could not find the requested resource") {
```

**Review by misraved — COMMENTED** [review]
@bigdatasourav please take a look at the review comments. Thanks!!

File: docs/tables/kubernetes_{custom_resource_singular_name}.md [hunk 12]
```diff
@@ -0,0 +1,446 @@
+# Table: kubernetes_{custom_resource_singular_name}
+
+Query data from the custom resource called `kubernetes_{custom_resource_singular_name}`, e.g., `kubernetes_certificate`, `kubernetes_capacityrequest`. A table is automatically created to represent each custom resource.
```
[the remark this answers is not in the document]
Comment (bigdatasourav) [reply docs/tables/kubernetes_{custom_resource_singular_name}.md:3]: no, here, capacityrequest is the singular name of the custom resource.

File: kubernetes/plugin.go (hunk 9 above)
Comment (bigdatasourav) [reply kubernetes/plugin.go:128]: This will execute at plugin load time, the user will not have any option here to specify a limit.

**Pull request merged**